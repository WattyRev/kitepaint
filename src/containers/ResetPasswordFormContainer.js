import React from "react";
import ResetPasswordForm from "../components/ResetPasswordForm";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { RESET_PASSWORD } from "../redux/actions";

export class ResetPasswordFormContainer extends React.Component {
  static propTypes = {
    /**
     * A unique identifier needed to differentiate instances of this componnet.
     */
    id: PropTypes.string.isRequired,
    /**
     * Indicates if the form is disabled.
     */
    isDisabled: PropTypes.bool,
    /**
     * A function called when the user clicks on the cancel button.
     */
    onCancel: PropTypes.func.isRequired,
    /**
     * A function called when the form is submitted.
     * Provided by redux.
     */
    onSubmit: PropTypes.func.isRequired
  };
  state = {
    /**
     * The user entered email address. Used on the registration and reset password forms.
     * @type {String}
     */
    email: "",
    /**
     * An error message to display when reset password fails.
     * @type {String}
     */
    resetPasswordErrorMessage: null,
    /**
     * Indicates that a reset password request has been sent. Should be set to true after submission
     * of the reset password form, that way we can display a success message to the user.
     * @type {Boolean}
     */
    resetPasswordSent: false,
    /**
     * The user entered username. Used on the registration, log in, and reset password forms.
     * @type {String}
     */
    username: ""
  };

  /**
   * Handles changes to the email field by storing it in state
   * @param  {String} email The new email address
   */
  handleEmailChange = email => {
    this.setState({
      email
    });
  };

  /**
   * Handles changes to the username field by storing it in state
   * @param  {String} username The new username
   */
  handleUsernameChange = username => {
    this.setState({
      username
    });
  };

  /**
   * Handles the password reset form submission.
   */
  handleSubmit = () => {
    return this.props
      .onSubmit(this.state.username, this.state.email)
      .then(() => {
        this.setState({
          resetPasswordSent: true
        });
      })
      .catch(error => {
        this.setState({
          resetPasswordErrorMessage: error
        });
      });
  };

  render() {
    return (
      <ResetPasswordForm
        email={this.state.email}
        errorMessage={this.state.resetPasswordErrorMessage}
        id={this.props.id}
        isDisabled={this.props.isDisabled}
        onCancel={this.props.onCancel}
        onEmailChange={this.handleEmailChange}
        onSubmit={this.handleSubmit}
        onUsernameChange={this.handleUsernameChange}
        showSuccessMessage={this.state.resetPasswordSent}
        username={this.state.username}
      />
    );
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  onSubmit: RESET_PASSWORD
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ResetPasswordFormContainer);
