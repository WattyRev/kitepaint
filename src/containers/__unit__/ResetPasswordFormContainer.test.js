import React from "react";
import { shallow } from "enzyme";
import { ResetPasswordFormContainer } from "../ResetPasswordFormContainer";

describe("ResetPasswordFormContainer", () => {
  let props;
  beforeEach(() => {
    props = {
      id: "abc",
      onSubmit: jest.fn(),
      onCancel: jest.fn()
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<ResetPasswordFormContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  describe("#handleUsernameChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new ResetPasswordFormContainer(props);
      subject.setState = jest.fn();
      subject.handleUsernameChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        username: "boogers",
        errorMessage: null
      });
    });
  });
  describe("#handleEmailChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new ResetPasswordFormContainer(props);
      subject.setState = jest.fn();
      subject.handleEmailChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        email: "boogers",
        errorMessage: null
      });
    });
  });
  describe("#handleSubmit", () => {
    it("maintains the pendingRequest state value", () => {
      expect.assertions(4);
      props.onSubmit.mockResolvedValue();
      const subject = new ResetPasswordFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false,
          resetPasswordSent: true
        });
      });
      return promise;
    });
    it("sets the error message if the call fails", () => {
      expect.assertions(4);
      props.onSubmit.mockRejectedValue("fuck you");
      const subject = new ResetPasswordFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false,
          errorMessage: "fuck you"
        });
      });
      return promise;
    });
  });
});
