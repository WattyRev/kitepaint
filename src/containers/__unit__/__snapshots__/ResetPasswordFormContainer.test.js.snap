// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordFormContainer renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResetPasswordFormContainer
    id="abc"
    onCancel={[MockFunction]}
    onSubmit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "email": "",
      "errorMessage": null,
      "id": "abc",
      "isDisabled": false,
      "onCancel": [MockFunction],
      "onEmailChange": [Function],
      "onSubmit": [Function],
      "onUsernameChange": [Function],
      "showSuccessMessage": false,
      "username": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "email": "",
        "errorMessage": null,
        "id": "abc",
        "isDisabled": false,
        "onCancel": [MockFunction],
        "onEmailChange": [Function],
        "onSubmit": [Function],
        "onUsernameChange": [Function],
        "showSuccessMessage": false,
        "username": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
